presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java

```diff
@@ -35,6 +35,37 @@
 import java.util.Map;
 import java.util.Optional;
 
+import static com.facebook.presto.sql.planner.plan.Assignments.identity;
+
+/**
+ * Transforms:
+ *
+ *  <pre>
+ *  Project(x = e1, y = e2)
+ *    Exchange()
+ *      Source(a, b, c)
+ *  </pre>
+ *
+ *  to:
+ *
+ *  <pre>
+ *  Exchange()
+ *    Project(x = e1, y = e2)
+ *      Source(a, b, c)
+ *  </pre>
+ *
+ *  Or if Exchange needs symbols from Source for partitioning or as hash symbol to:
+ *
+ *  <pre>
+ *  Project(x, y)
+ *    Exchange()
+ *      Project(x = e1, y = e2, a)
+ *        Source(a, b, c)
+ *  </pre>
+ *
+ *
+ *  To avoid looping this optimizer will not be fired if upper Project contains just symbol references.
```

Comment:
Discussed that it is about need for projection above Exchange node sometimes.

This is the issue: https://github.com/prestodb/presto/issues/7307